pandas/core/groupby/base.py

```diff
@@ -93,19 +94,21 @@ def _gotitem(self, key, ndim, subset=None):
                                    'idxmax'])
 
 
-def whitelist_method_generator(base, klass, whitelist):
+def whitelist_method_generator(base: 'Type[GroupBy]',
+                               klass: 'Union[Type[DataFrame], Type[Series]]',
```

Comment:
This should probably be a TypeVar instead of a Union (i.e. we parametrize on one or the other of these, but don't use interchangeably within the function).

This is going to be a common idiom - could you add `FrameOrSeries = TypeVar('FrameOrSeries', ABCSeries, ABCDataFrame)` to pandas._typing and use that here instead (open to other naming suggestions)

Reply:
I don't understand TypeVars all too well, but my understanding is it it ensures that the same type comes out as came in. So, if a Series came in, and an object of the same TypeVar comes out, that means the returnm type is a Series (and not DataFrame). In that case, it will not matter in this case if we use Union or TypeVar, isn't that true?

Not opposed to using TypeVar, just trying to understand their use cases.

Reply:
Yea you've got that right! This comment might also be helpful

https://github.com/pandas-dev/pandas/issues/26453#issuecomment-497025873

So I don't think will make a difference here - just looking to get some kind of consistency in place as our annotations are still relatively immature

Reply:
Ok. It will be actually be more concise to use that TypeVar here (fewer type-specific imports), and a ``TypeVar('FrameOrSeries', Series, DataFrame)``  would make sense for Pandas in general.

In Pycharm ``TypeVar('FrameOrSeries', ABCSeries, ABCDataFrame)`` doesn't give the correct completions, while ``TypeVar('FrameOrSeries', Series, DataFrame)`` does give the right ones. any specific reason to use ABCSeries and ABCDataFrame over the actual classes?

I'm also thinking that the best version would be actually  ``TypeVar('FrameOrSeries', NDFRame, Series, DataFrame)`, just to be able to add type relevant hints to the NDFrame class. Any thoughts on this?

Reply:
ABCSeries and ABCDataFrame are supposed to circumvent issues with import cycles. Unfortunately I don't use PyCharm so I can't speak to their support for those w.r.t. code completion.

I also don't think NDFrame should be a part of that as it's not a type that is used directly (i.e. typically only used as a superclass)

Reply:
> ABCSeries and ABCDataFrame are supposed to circumvent issues with import cycles.

To clarify though, if you see a better way of doing this certainly open to suggestions. I can't claim total expertise in our comprehensive usage of the ABC classes. @jreback

Reply:
@WillAyd, do you use VSCode? Do you get correct type hints from TypeVars?

Reply:
No. Haven't used annotations for code completion so any insights you can offer there would be appreciated